storage/mysql/tree_storage_test.go

```diff
@@ -91,23 +139,542 @@ func TestNodeRoundTrip(t *testing.T) {
+	err := tx.QueueLeaves(leaves)
+
+	if err != nil {
+		t.Errorf("Failed to queue leaves: %v", err)
+		t.FailNow()
+	}
 
+	leaves2 := createTestLeaves(5, 12)
+	err = tx.QueueLeaves(leaves2)
+
+	if err == nil {
+		t.Error("Allowed duplicate leaves to be inserted")
+		t.FailNow()
+	}
+
+	if !strings.Contains(err.Error(), "Duplicate") {
+		t.Errorf("Got the wrong type of error: %v", err)
+	}
 }
 
-func createTestDB() {
+func TestQueueLeaves(t *testing.T) {
+	logID := createLogID("TestQueueLeaves")
+	db := prepareTestDB(logID, t)
+	defer db.Close()
+	s := prepareTestStorage(logID, t)
+  tx := beginTx(s, t)
+
+	leavesToInsert := 5
+	leaves := createTestLeaves(leavesToInsert, 20)
+	err := tx.QueueLeaves(leaves)
+
+	if err != nil {
+		t.Errorf("Failed to queue leaves: %v", err)
+		t.FailNow()
+	}
+
+	commit(tx, t)
+
+	// Should see the leaves in the database. There is no API to read from the
+	// unsequenced data.
+	var count int
+	err = db.QueryRow("SELECT COUNT(*) FROM Unsequenced WHERE TreeID=?", logID.logID.TreeID).Scan(&count)
+
+	if err != nil {
+		t.Fatalf("Could not query row count")
+	}
+
+	if count != leavesToInsert {
+		t.Errorf("Expected %d unsequenced rows but got: %d", leavesToInsert, count)
+	}
+}
+
+func TestDequeueLeavesNoneQueued(t *testing.T) {
+	logID := createLogID("TestDequeueLeavesNoneQueued")
+	db := prepareTestDB(logID, t)
+	defer db.Close()
+	s := prepareTestStorage(logID, t)
+	tx := beginTx(s, t)
+
+	leaves, err := tx.DequeueLeaves(999)
+
+	if err != nil {
+		t.Fatalf("Didn't expect an error on dequeue with no work to be done: %v", err)
+	}
+
+	if len(leaves) > 0 {
+		t.Errorf("Expected nothing to be dequeued but we got %d leaves", len(leaves))
+	}
+}
+
+func TestDequeueLeaves(t *testing.T) {
+	logID := createLogID("TestDequeueLeaves")
+	db := prepareTestDB(logID, t)
+	defer db.Close()
+	s := prepareTestStorage(logID, t)
+	tx := beginTx(s, t)
+
+	leavesToInsert := 5
+	leaves := createTestLeaves(leavesToInsert, 20)
+	err := tx.QueueLeaves(leaves)
+
+	if err != nil {
+		t.Errorf("Failed to queue leaves: %v", err)
+		t.FailNow()
+	}
+
+	commit(tx, t)
+
+	// Now try to dequeue them
+	tx = beginTx(s, t)
+	leaves2, err := tx.DequeueLeaves(99)
+
+	if err != nil {
+		t.Fatalf("Failed to dequeue leaves: %v", err)
+	}
+
+	if len(leaves2) != leavesToInsert {
+		t.Fatalf("Dequeued %d leaves but expected to get %d", len(leaves), leavesToInsert)
+	}
+
+	ensureAllLeafHashesDistinct(leaves2, t)
+
+	tx.Commit()
+
+	// If we dequeue again then we should now get nothing
+	tx = beginTx(s, t)
+	leaves2, err = tx.DequeueLeaves(99)
+
+	if err != nil {
+		t.Fatalf("Failed to dequeue leaves (second time): %v", err)
+	}
+
+	if len(leaves2) != 0 {
+		t.Fatalf("Dequeued %d leaves but expected to get none", len(leaves))
+	}
+}
+
+func TestDequeueLeavesTwoBatches(t *testing.T) {
+	logID := createLogID("TestDequeueLeaves")
+	db := prepareTestDB(logID, t)
+	defer db.Close()
+	s := prepareTestStorage(logID, t)
+	tx := beginTx(s, t)
+
+	leavesToInsert := 5
```

Comment:
more consts? (probably more elsewhere, but I'll stop calling them out to cut down on noise)